### PR TITLE
Added ExponentialRK, IMEXMultistep and Linear Solvers

### PR DESCRIPTION
**Pull request by ParamThakkar123**
## Checklist

- [X] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [X] Any new documentation only uses public API
  
## Additional context

solves a part of #2177

### CODE REVIEW

File: lib/OrdinaryDiffEqLinear/src/OrdinaryDiffEqLinear.jl [hunk 1]
```diff
@@ -0,0 +1,24 @@
+module OrdinaryDiffEqLinear
+
+import OrdinaryDiffEq: alg_order, alg_extrapolates, dt_required, OrdinaryDiffEqLinearExponentialAlgorithm,
+                       OrdinaryDiffEqAdaptiveAlgorithm, OrdinaryDiffEqAlgorithm,
+                       OrdinaryDiffEqExponentialAlgorithm,
+                       OrdinaryDiffEqMutableCache, @cache, alg_cache, OrdinaryDiffEqConstantCache,
+                       initialize!, perform_step!, @unpack, unwrap_alg, calculate_residuals!,
+                       _vec
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLinear/src/OrdinaryDiffEqLinear.jl:8]:
```suggestion
                       _vec, isdtchangeable
```

File: lib/OrdinaryDiffEqExponentialRK/src/OrdinaryDiffEqExponentialRK.jl [hunk 2]
```diff
@@ -0,0 +1,23 @@
+module OrdinaryDiffEqExponentialRK
+
+import OrdinaryDiffEq: alg_order, alg_adaptive_order, ismultistep, OrdinaryDiffEqExponentialAlgorithm,
+                       _unwrap_val, OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       build_jac_config, UJacobianWrapper, @cache, alg_cache, UDerivativeWrapper,
+                       initialize!, perform_step!, @unpack, unwrap_alg, calc_J, calc_J!,
+                       OrdinaryDiffEqAdaptiveExponentialAlgorithm, CompositeAlgorithm,
+                       ExponentialAlgorithm
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqExponentialRK/src/OrdinaryDiffEqExponentialRK.jl:8]:
```suggestion
                       ExponentialAlgorithm, fsal_typeof
```

File: lib/OrdinaryDiffEqExponentialRK/src/OrdinaryDiffEqExponentialRK.jl [hunk 3]
```diff
@@ -0,0 +1,23 @@
+module OrdinaryDiffEqExponentialRK
+
+import OrdinaryDiffEq: alg_order, alg_adaptive_order, ismultistep, OrdinaryDiffEqExponentialAlgorithm,
+                       _unwrap_val, OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       build_jac_config, UJacobianWrapper, @cache, alg_cache, UDerivativeWrapper,
+                       initialize!, perform_step!, @unpack, unwrap_alg, calc_J, calc_J!,
+                       OrdinaryDiffEqAdaptiveExponentialAlgorithm, CompositeAlgorithm,
+                       ExponentialAlgorithm, fsal_typeof
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqExponentialRK/src/OrdinaryDiffEqExponentialRK.jl:8]:
```suggestion
                       ExponentialAlgorithm, fsal_typeof, isdtchangeable
```

File: lib/OrdinaryDiffEqLinear/src/OrdinaryDiffEqLinear.jl [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import OrdinaryDiffEq: alg_order, alg_extrapolates, dt_required, OrdinaryDiffEqLinearExponentialAlgorithm,
+                       OrdinaryDiffEqAdaptiveAlgorithm, OrdinaryDiffEqAlgorithm,
+                       OrdinaryDiffEqExponentialAlgorithm,
+                       OrdinaryDiffEqMutableCache, @cache, alg_cache, OrdinaryDiffEqConstantCache,
+                       initialize!, perform_step!, @unpack, unwrap_alg, calculate_residuals!,
+                       _vec, isdtchangeable
+using LinearAlgebra: mul!
+using DiffEqBase
+using SciMLOperators: AbstractSciMLOperator
+using ExponentialUtilities
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLinear/src/OrdinaryDiffEqLinear.jl:12]:
```suggestion
using ExponentialUtilities
using RecursiveArrayTools
```

File: lib/OrdinaryDiffEqLinear/src/OrdinaryDiffEqLinear.jl [hunk 5]
```diff
@@ -0,0 +1,25 @@
+module OrdinaryDiffEqLinear
+
+import OrdinaryDiffEq: alg_order, alg_extrapolates, dt_required, OrdinaryDiffEqLinearExponentialAlgorithm,
+                       OrdinaryDiffEqAdaptiveAlgorithm, OrdinaryDiffEqAlgorithm,
+                       OrdinaryDiffEqExponentialAlgorithm,
+                       OrdinaryDiffEqMutableCache, @cache, alg_cache, OrdinaryDiffEqConstantCache,
+                       initialize!, perform_step!, @unpack, unwrap_alg, calculate_residuals!,
+                       _vec, isdtchangeable
+using LinearAlgebra: mul!
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLinear/src/OrdinaryDiffEqLinear.jl:9]:
```suggestion
using LinearAlgebra: mul!, I
```

File: lib/OrdinaryDiffEqExponentialRK/src/OrdinaryDiffEqExponentialRK.jl [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+                       build_jac_config, UJacobianWrapper, @cache, alg_cache, UDerivativeWrapper,
+                       initialize!, perform_step!, @unpack, unwrap_alg, calc_J, calc_J!,
+                       OrdinaryDiffEqAdaptiveExponentialAlgorithm, CompositeAlgorithm,
+                       ExponentialAlgorithm, fsal_typeof, isdtchangeable
+using RecursiveArrayTools
+using MuladdMacro, FastBroadcast
+using LinearAlgebra: axpy!, mul!
+using DiffEqBase, SciMLBase
+using ExponentialUtilities
+
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqExponentialRK/src/OrdinaryDiffEqExponentialRK.jl:14]:
```suggestion
using ExponentialUtilities
import RecursiveArrayTools: recursivecopy!

```

File: lib/OrdinaryDiffEqExponentialRK/src/OrdinaryDiffEqExponentialRK.jl [hunk 7]
```diff
@@ -0,0 +1,24 @@
+module OrdinaryDiffEqExponentialRK
+
+import OrdinaryDiffEq: alg_order, alg_adaptive_order, ismultistep, OrdinaryDiffEqExponentialAlgorithm,
+                       _unwrap_val, OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache,
+                       build_jac_config, UJacobianWrapper, @cache, alg_cache, UDerivativeWrapper,
+                       initialize!, perform_step!, @unpack, unwrap_alg, calc_J, calc_J!,
+                       OrdinaryDiffEqAdaptiveExponentialAlgorithm, CompositeAlgorithm,
+                       ExponentialAlgorithm, fsal_typeof, isdtchangeable
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqExponentialRK/src/OrdinaryDiffEqExponentialRK.jl:8]:
```suggestion
                       ExponentialAlgorithm, fsal_typeof, isdtchangeable, calculate_residuals, calculate_residuals!
```

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas What all things are yet to be done now ??

**Comment by ChrisRackauckas** [comment]
I'm going to change up the CI https://github.com/SciML/OrdinaryDiffEq.jl/pull/2311 so that it actually tests the submodules in isolation. I think this will likely find a few more issues with the individual subpackages that wasn't noticed when always running them with the global OrdinaryDiffEq. I might enjoy some help with that. Also we need to split differentiation, nonlinearsolve, and core and that's started in https://github.com/SciML/OrdinaryDiffEq.jl/pull/2309 . These two pieces kind of need to go together, since all of the `using OrdinaryDiffEq` should become `using OrdinaryDiffEqCore`, so that in theory these can all be used in isolation. So there's still some more to do before we're really ready to release. 

But if you want to claim your victory payout, I think you completed what I had originally outlined about splitting the solvers out, and I'd be willing to say that's completed and start that process.

**Comment by ParamThakkar123** [comment]
> I'm going to change up the CI #2311 so that it actually tests the submodules in isolation. I think this will likely find a few more issues with the individual subpackages that wasn't noticed when always running them with the global OrdinaryDiffEq. I might enjoy some help with that. Also we need to split differentiation, nonlinearsolve, and core and that's started in #2309 . These two pieces kind of need to go together, since all of the `using OrdinaryDiffEq` should become `using OrdinaryDiffEqCore`, so that in theory these can all be used in isolation. So there's still some more to do before we're really ready to release.
> 
> But if you want to claim your victory payout, I think you completed what I had originally outlined about splitting the solvers out, and I'd be willing to say that's completed and start that process.

Ok. I would also love to be a part of this further and help with the further processes that are yet to be done

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas You can assign me some tasks where you think I can help

**Comment by ParamThakkar123** [comment]
> But if you want to claim your victory payout, I think you completed what I had originally outlined about splitting the solvers out, and I'd be willing to say that's completed and start that process.

We can discuss this over the mail

**Comment by ChrisRackauckas** [comment]
> You can assign me some tasks where you think I can help

Thanks! I just merged https://github.com/SciML/OrdinaryDiffEq.jl/pull/2311 which has some failing tests to handle. You can see the last commit I made which was making progress towards it. With this change to the CI, it's now testing each of the submodules as a separate package. This means that, unlike before, the Project.toml per-package is now being checked to be correct for the test dependencies. That'll take some working through. If you do that, then I can do the OrdinaryDiffEqDifferentiation/OrdinaryDiffEqNonlinear/OrdinaryDiffEqCore change, and then there will be another phase of fixing the Project.tomls so that each package has the right dependencies everywhere. Once that's done, the split is complete!

**Comment by ParamThakkar123** [comment]
> > You can assign me some tasks where you think I can help
> 
> Thanks! I just merged #2311 which has some failing tests to handle. You can see the last commit I made which was making progress towards it. With this change to the CI, it's now testing each of the submodules as a separate package. This means that, unlike before, the Project.toml per-package is now being checked to be correct for the test dependencies. That'll take some working through. If you do that, then I can do the OrdinaryDiffEqDifferentiation/OrdinaryDiffEqNonlinear/OrdinaryDiffEqCore change, and then there will be another phase of fixing the Project.tomls so that each package has the right dependencies everywhere. Once that's done, the split is complete!

Ok I will take this part of adding the test dependencies to all the packages

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas just sent you an email on this : admin@chrisrackauckas.com

**Pull request merged**